src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java

```diff
@@ -598,127 +624,157 @@ private SocketAddress untrustedReceive(ByteBuffer dst) throws IOException {
-    {
-        assert readLock.isHeldByCurrentThread() && isBlocking();
-        SocketAddress sender = null;
+    void blockingReceive(DatagramPacket p, long nanos) throws IOException {
+        byte[] ba; int off, len;
+        synchronized (p) {
+            ba = p.getData();
+            off = p.getOffset();
+            len = DatagramPackets.getBufLength(p);
+        }
+
+        int n = -1;
+        InetSocketAddress sender = null;
+
+        readLock.lock();
         try {
-            SocketAddress remote = beginRead(true, false);
-            boolean connected = (remote != null);
+            ensureOpen();
+            if (!isBlocking())
+                throw new IllegalBlockingModeException();
 
-            // change socket to non-blocking
-            lockedConfigureBlocking(false);
             try {
-                long startNanos = System.nanoTime();
-                int n = receive(dst, connected);
-                while (n == IOStatus.UNAVAILABLE && isOpen()) {
-                    long remainingNanos = nanos - (System.nanoTime() - startNanos);
-                    if (remainingNanos <= 0) {
-                        throw new SocketTimeoutException("Receive timed out");
+                SocketAddress remote = beginRead(true, false);
+                configureSocketNonBlockingIfVirtualThread();
+                boolean connected = (remote != null);
+                if (nanos > 0) {
+                    // timed receive, change socket to non-blocking
+                    lockedConfigureBlocking(false);
+                    try {
+                        n = timedReceive(ba, off, len, nanos, connected);
+                    } finally {
+                        // restore socket to blocking mode
+                        tryLockedConfigureBlocking(true);
+                    }
+                } else {
+                    // untimed receive
+                    n = tryReceive(ba, off, len, connected);
+                    while (IOStatus.okayToRetry(n) && isOpen()) {
+                        park(Net.POLLIN);
+                        n = tryReceive(ba, off, len, connected);
                     }
-                    park(Net.POLLIN, remainingNanos);
-                    n = receive(dst, connected);
                 }
                 if (n >= 0) {
                     // sender address is in socket address buffer
                     sender = sourceSocketAddress();
                 }
-                return sender;
             } finally {
-                // restore socket to blocking mode (if channel is open)
-                tryLockedConfigureBlocking(true);
+                endRead(true, n >= 0);
             }
         } finally {
-            endRead(true, (sender != null));
+            readLock.unlock();
+        }
+
+        // set datagram length and sender address
+        assert n >= 0 && sender != null;
+        synchronized (p) {
+            DatagramPackets.setLength(p, n);
+            p.setSocketAddress(sender);
```

Comment:
Hmmm... For integrity it might be better to call the public  `DatagramPacket::setData(byte[] buf, int offset, int length)` method here now. The additional advantage is that the private access to  `DatagramPackets.setLength(p, n);` will no longer be needed.

Reply:
DatagramPackets.setLength is not really new, it's just moved. There's a flaw in DatagramPacket that forces its usage -  the details are in JDK-8232817.